src/develop/masks/brush.c

```diff
@@ -2539,67 +2524,58 @@ static int dt_brush_get_source_area(dt_iop_module_t *module, dt_dev_pixelpipe_io
     ymin = fminf(yy, ymin);
     ymax = fmaxf(yy, ymax);
   }
+  *x_min = xmin;
+  *x_max = xmax;
+  *y_min = ymin;
+  *y_max = ymax;
+}
 
-  dt_free_align(points);
-  dt_free_align(border);
+static void _brush_bounding_box(const float *const points, const float *const border, const int nb_corner, const int num_points, const int num_borders,
+                                int *width, int *height, int *posx, int *posy)
+{
+  float xmin = FLT_MAX, xmax = FLT_MIN, ymin = FLT_MAX, ymax = FLT_MIN;
+  _brush_bounding_box_raw(points, border, nb_corner, num_points, num_borders, &xmin, &xmax, &ymin, &ymax);
   *height = ymax - ymin + 4;
   *width = xmax - xmin + 4;
   *posx = xmin - 2;
   *posy = ymin - 2;
-  return 1;
 }
 
-static int dt_brush_get_area(dt_iop_module_t *module, dt_dev_pixelpipe_iop_t *piece, dt_masks_form_t *form,
-                             int *width, int *height, int *posx, int *posy)
+static int _brush_get_area(dt_iop_module_t *module, dt_dev_pixelpipe_iop_t *piece,
+                           dt_masks_form_t *form, int *width, int *height, int *posx, int *posy, int get_source)
 {
   if(!module) return 0;
   // we get buffers for all points
   float *points = NULL, *border = NULL;
   int points_count, border_count;
   if(!_brush_get_points_border(module->dev, form, module->iop_order, DT_DEV_TRANSFORM_DIR_BACK_INCL, piece->pipe, &points, &points_count,
-                               &border, &border_count, NULL, NULL, 0))
+                               &border, &border_count, NULL, NULL, get_source))
   {
     dt_free_align(points);
     dt_free_align(border);
     return 0;
   }
 
-  // now we want to find the area, so we search min/max points
-  float xmin = 0.0f, xmax = 0.0f, ymin = 0.0f, ymax = 0.0f;
-  xmin = ymin = FLT_MAX;
-  xmax = ymax = FLT_MIN;
   const guint nb_corner = g_list_length(form->points);
-  for(int i = nb_corner * 3; i < border_count; i++)
-  {
-    // we look at the borders
-    const float xx = border[i * 2];
-    const float yy = border[i * 2 + 1];
-    xmin = fminf(xx, xmin);
-    xmax = fmaxf(xx, xmax);
-    ymin = fminf(yy, ymin);
-    ymax = fmaxf(yy, ymax);
-  }
-  for(int i = nb_corner * 3; i < points_count; i++)
-  {
-    // we look at the brush too
-    const float xx = points[i * 2];
-    const float yy = points[i * 2 + 1];
-    xmin = fminf(xx, xmin);
-    xmax = fmaxf(xx, xmax);
-    ymin = fminf(yy, ymin);
-    ymax = fmaxf(yy, ymax);
-  }
+  _brush_bounding_box(points, border, nb_corner, points_count, border_count, width, height, posx, posy);
 
   dt_free_align(points);
   dt_free_align(border);
-
-  *height = ymax - ymin + 4;
-  *width = xmax - xmin + 4;
-  *posx = xmin - 2;
-  *posy = ymin - 2;
   return 1;
 }
 
+static int dt_brush_get_source_area(dt_iop_module_t *module, dt_dev_pixelpipe_iop_t *piece,
+                                    dt_masks_form_t *form, int *width, int *height, int *posx, int *posy)
```

Comment:
Looks like this gets called only once? Why not refactor that location to call _brush_get_area directly (with the area=1 parameter added). I find tracing all these intermediate pass-through functions does take mental space. (I know it probably gets inlined, so speedwise wouldn't make a difference).

Also, static (local) functions shouldn't be dt_? (might as well fix that, while you are touching these).

Thanks for your amazingly continuing clean-up efforts!

Reply:
Both of your comments really apply to the majority of the `dt_*` functions in the $SHAPE.c files, due to the poor software engineering there (those files are treated like header files, included in masks.c, and surrounded by a compiler directive that I suspect is to make clang shut up about the practice).  I'd rather not make your suggested refactor because it would just need to be undone on an overall cleanup such as splitting the separate .c files into separate compilation units.

An even better approach to the whole thing would be to have a table of function pointers for each kind of shape (with a pointer to the appropriate table in each shape instance), to eliminate the long if-else chains for each operation and make it simpler to add a new shape such as rectangle.

Reply:
Clearly I'd have to look at this much longer to understand what's really going on. My fear just is that adding another layer of pass-through functions makes that even more so for the next person who takes a look, making an eventual cleanup less likely. But I don't blame you for not wanting to tackle it now.

Reply:
The eventual cleanup has just been done :-)
#8290